### PR TITLE
450 sizeof() operator

### LINKED ISSUE #450: Add support for "sizeof()" operator
The parser is currently failing to handle the "sizeof()" operator correctly. Here's an example that illustrates the issue:

`int main() {
    return sizeof(int);
}`

From the standard is Section 6.5.3.4 The sizeof and _Alignof operators
I will be working on this issue.
Related to #207

### PR DESCRIPTION
**Pull request by abrahamFerga**
This pull request closes #450, there are a few cases where sizeof() is failing (see the Integration Tests).

Let me know if any change is required or if I should wait for #453 (next on my list).

### CODE REVIEW

File: Cesium.CodeGen/Contexts/TranslationUnitContext.cs [hunk 1]
```diff
@@ -168,7 +168,8 @@ internal IType ResolveType(IType type)
         return type;
     }
 
-    internal TypeReference? GetTypeReference(IGeneratedType type) => _generatedTypes.GetValueOrDefault(type);
+    internal TypeReference? GetTypeReference(IGeneratedType type) => _generatedTypes.GetValueOrDefault(type) ??
+                                                                     _generatedTypes.FirstOrDefault(t => t.Key.Identifier == type.Identifier).Value;
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Contexts/TranslationUnitContext.cs]: I don't believe this change is correct. The chance to mess something up (like `struct foo` vs `typedef foo struct { … }`) is very high.

I think that the type _should_ be emitted at this point.

For now, let's roll this statement back and fall back to `typedef` in a test, and figure the rest out in #453.

File: Cesium.CodeGen/Ir/Expressions/SizeOfOperatorExpression.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            ({ }, { }) => LocalDeclarationInfo.Of(specifiers, abstractDeclarator).Type,
+            ({ }, null) => LocalDeclarationInfo.Of(specifiers, (Declarator?)null).Type
+        };
+    }
+
+    public IExpression Lower(IDeclarationScope scope)
+    {
+        if (_type is NamedType namedType)
+        {
+            var typeResolved = new IdentifierExpression(namedType.TypeName).Resolve(scope).GetValueType();
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/SizeOfOperatorExpression.cs:32]: Currently, this definition, and perhaps even parsing for this, is incorrect.

We want to support both `sizeof(variable)` and `sizeof(type)`, right?

But so far the parser only supports `sizeof(type_name)`, and thus it _accidentally_ supports _variables_ that _look like type names_, which cannot be right 😅 

This leads to further problems, one of which is now demonstrated in a failed `typedef` test I pushed into the branch. In my opinion, it looks like you are trying to save the incorrect definition by resolving a `NamedType` as a variable name.

This is of course move in the right direction, but let's work a little bit more on the definition of this: perhaps let's introduce two different `sizeof` variants for types and variables? (and then accept just an `identifier` and see how it resolves in the lower/emit time)

Note I'm okay to disable the (currently) failing tests and merge this while opening a follow-up issue on this, but probably I'd like to disable the support for _variables_ completely beforehand. I am open for discussion, though. If you are very eager to merge this as is, I'm ready to negotiate somewhat :)

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

File: Cesium.CodeGen/Contexts/TranslationUnitContext.cs (hunk 1 above)
Comment (ForNeVeR) [reply Cesium.CodeGen/Contexts/TranslationUnitContext.cs]: Note I have rolled it back and pushed a commit to demonstrate an issue with this approach. Please tell if you are uncomfortable with such a move from me :)

Comment (abrahamFerga) [reply Cesium.CodeGen/Contexts/TranslationUnitContext.cs]: I knew something was wrong with this but it was the only way I found to get the type. I will work in #453 and we won't need this anymore.
No worries, I'm here to learn, any feedback is well appreciated.

File: Cesium.CodeGen/Ir/Expressions/SizeOfOperatorExpression.cs (hunk 2 above)
Comment (abrahamFerga) [reply Cesium.CodeGen/Ir/Expressions/SizeOfOperatorExpression.cs:32]: At first I had two different expressions for this but somehow it worked having only one of them as `sizeof(type_name)`, sure I will split both scenarios and I will update this pull request along the other issue mentioned above.

Thank you for the feedback!

**Comment by abrahamFerga** [comment]
Hey there,

I've done some work to split up the cases for sizeof(Identifier) and sizeof(TypeName). It's working well in most situations, but there's a hiccup with structs since we haven't figured out a way to resolve them by identifier yet.

I'll be honest, my knowledge of C is somewhat limited, and the deeper I dive into it, the more peculiar things get. There are a few edge cases that this PR doesn't handle, like sizeof(5*5). However, I think these changes, while interesting, might not be necessary for the foreseeable future.

Cheers.

**Comment by kant2002** [comment]
I personally believe that if this is half-done feature which does not cover all cases that's great addition. We work in small steps on such things. Because usually it's too draining work on full spec. If you think that you never introduce any "wrong" behaviour, then this should be fine.

Also, you doing great job. I really like to see all these additions to Cesium! Keep going!

**Review by ForNeVeR — APPROVED** [review]
Thanks!

**Pull request merged**